sgr_agent_core/agents/tool_calling_agent.py

```diff
@@ -106,3 +88,23 @@ async def _action_phase(self, tool: BaseTool) -> str:
         self.streaming_generator.add_chunk_from_str(f"{result}\n")
         self._log_tool_execution(tool, result)
         return result
+
+
+class ResearchToolCallingAgent(ToolCallingAgent):
+    async def _prepare_tools(self) -> list[ChatCompletionFunctionToolParam]:
+        """Prepare available tools for the current agent state and progress."""
+        tools = set(self.toolkit)
+        if self._context.iteration >= self.config.execution.max_iterations:
+            tools = {
```

Comment:
how about
````|
tools.intersection_update({
                CreateReportTool,
                FinalAnswerTool,
            })
            ```
            ?
Because these tools may be not included in configuration (global or specific agent).

Reply:
good question.

One of concepts im trying to implement here is: we have more general purpose agents not depending on tools and have Deep Search versions because of its our historical  primary bench/usage case.

So, research versions include some prebuilt logic/funcs. Maybe it will be a good point to add this tooling as default kit, but what if user wants add custom implementation of search tool?